TypeCobol.LanguageServer/TypeCobolServer.cs

```diff
@@ -544,6 +545,17 @@ protected override Hover OnHover(TextDocumentPosition parameters)
                         }
                     }
                     break;
+                default:
+                    // Use userFilterToken to find the variable under mouse cursor.
+                    if (userFilterToken != null)
+                    {
+                        var candidates = matchingNode.SymbolTable.GetVariablesExplicit(new URI(userFilterToken.SourceText)).ToList();
```

Comment:
All information about Tokens<->SymbolInformation<->StorageArea<->Node are stored in CodeElement and Node.
See 
`CodeElement.SymbolInformationForTokens`
Then when you have the SymbolInformation you can iterate list:
- `CodeElement.StorageAreaReads`
- `CodeElement.StorageAreaWrites`

To check if a StorageArea contains this SymbolInformation 

Then with the StorageArea you can retrieve the corresponding DataDefinition already resolved by CrossChecker with
`node.GetDataDefinitionFromStorageAreaDictionary()`.

If one element is not found (eg. CodeElement.SymbolInformationForTokens doesn't contains the token, it means the token is not a SymbolInformation) and you can stop (there is no variable to hover).

Reply:
A quick code to illustrate:
```c#
// Use userFilterToken to find the variable under mouse cursor.
                    if (userFilterToken != null)
                    {

                        if(matchingNode.CodeElement.SymbolInformationForTokens.TryGetValue(userFilterToken, out var symbolInformation) )
                        {
                            //Search if a StorageArea is related to this SymbolInformation
                            StorageArea storageArea = null;
                            foreach (var tempStorageArea in matchingNode.CodeElement.StorageAreaReads)
                            {
                                if (tempStorageArea.SymbolReference == symbolInformation)
                                {
                                    storageArea = tempStorageArea;
                                    break;
                                }
                            }
                            foreach (var tempStorageArea in matchingNode.CodeElement.StorageAreaWrites)
                            {
                                if (tempStorageArea.StorageArea?.SymbolReference == symbolInformation)
                                {
                                    storageArea = tempStorageArea.StorageArea;
                                    break;
                                }
                            }
                            //End of search

                            if (storageArea != null)
                            {
                                //Get the cached DataDefinition for this StorageArea
                                var dataDef = matchingNode.GetDataDefinitionFromStorageAreaDictionary(storageArea);

                                if (dataDef != null)
                                {
                                    //Display the DataDefinition
                                    message = dataDef.ToString();
                                }
                            }

                        }
                    }
```

Reply:
Beware GetHashCode() method is missing in Token class.
So `matchingNode.CodeElement.SymbolInformationForTokens.TryGetValue(userFilterToken, out var symbolInformation) ` will not return anything.
A quick hash code for Token:
```C#
public override int GetHashCode()
        {
            return this.Type.GetHashCode() * 37
                   + this.Column * 37
                   + this.Length * 37 
                   + this.Line *37
                   + this.Text.GetHashCode();
        }
```
But I need to check if this hashCode can be improved.